### ISSUE TITLE
Dashboard Authentication not working

### ISSUE THREAD

**Issue by alain-baxter** [OpenUnison/openunison-k8s-login-activedirectory#57]
I've installed this to my cluster using the following steps:
* Installed Nginx Ingress and Kubernetes Dashboard
* Created config directory with input.props configured and trusted-ad.pem
* Created secret directory with input.props configured
* Ran the install command
* Verified OpenUnison comes up
```
# kubectl get pods -n openunison
NAME                                     READY   STATUS      RESTARTS   AGE
check-certs-orchestra-1580868000-52zlj   0/1     Completed   0          37h
check-certs-orchestra-1580954400-ltlg5   0/1     Completed   0          13h
openunison-operator-57cb6966c4-g92dr     1/1     Running     0          42h
openunison-orchestra-c97c7c99b-hzg7t     1/1     Running     0          42h
```
* Got the ou-ca.pem content with `kubectl get secret ou-tls-certificate -n openunison -o=jsonpath='{..tls\.crt}' | base64 --decode`
* Created /etc/kubernetes/pki/ou-ca.pem with the cert
* Configured kube-apiserver.yaml with the oicd server flags

I can login to the Auth Portal using AD Credentials, and see my roles from AD

I tested kubectl authentication successfully

I added cluster-admin permission to one of my groups and was able to successfully get all objects via kubectl.

**Issue:**
When I click on the "Kubernetes Dashboard" tile on the Home Page, I get forwarded to the Dashboard deployed to my cluster as expected, however I get loads of "Unauthorized" notifications, and can see nothing in the cluster.

When I inspect the requests, I can see that there is no `Authorization: Bearer <token>` header present,

**Versions:**
Cluster: Kubernetes v1.16.4
Chrome: 79.0.3945.130 (Official Build) (64-bit)
openunison-k8s-login-activedirectory: tremolosecurity/openunison-k8s-login-activedirectory@sha256:b9f5ac8d0212f7675bed5031b0fdcdd7e4cec79add9f050193c9b94586f2392a
Kubernetes Dashboard: kubernetesui/dashboard:v2.0.0-beta8@sha256:fc90baec4fb62b809051a3227e71266c0427240685139bbd5673282715924ea7

**Comment by mlbiam** [comment]
How are you inspecting the traffic?

**Comment by mlbiam** [comment]
I just ran a dry run with a fresh install and everything worked as planned.  Is traffic for the host you set for `K8S_DASHBOARD_HOST` going to OpenUnison or to its own Ingress?  If you look in the OpenUnison logs you should see log entries like:

```
[2020-02-07 01:11:05,754][XNIO-1 task-1] INFO  AccessLog - [AzSuccess] - k8s - https://k8sdb.apps.192-168-159-168.nip.io/api/v1/pod/openunison/openunison-orchestra-c97c7c99b-5d7xm/container - uid=mmosley,ou=shadow,o=Tremolo - oauth2bearer [10.244.0.4] - [f4a39d8a8927b7d175b75568e7f3f9918e9c02a9d]
```

**Comment by alain-baxter** [comment]
I was inspecting the traffic using Chrome Developer Tools.

The traffic for `K8S_DASHBOARD_HOST` is going through OpenUnison, since it is proxying traffic to the Dashboard using it's own Ingress Configuration.

I have the following showing up in my OpenUnison logs:
```
[2020-02-07 19:47:57,174][XNIO-1 task-11] INFO  AccessLog - [AzSuccess] - k8sIdp - https://k8s-auth-prototype.corp.cira.ca/auth/idp/k8sIdp/completefed - uid=alain.baxter,ou=shadow,o=Tremolo -  [192.168.217.220] - [f42f2a5275c48753c071d2332fdebd1bfc538ebeb]
[2020-02-07 19:47:57,238][XNIO-1 task-10] INFO  AccessLog - SRCH op=322 con=321 base='uid=alain.baxter,ou=shadow,o=Tremolo' filter='(objectClass=*)' scope='0' attribs=''
[2020-02-07 19:47:57,249][XNIO-1 task-10] INFO  AccessLog - RESULT op=322 con=321 result=0 time=11
[2020-02-07 19:47:57,251][XNIO-1 task-10] INFO  AccessLog - SRCH op=323 con=322 base='uid=alain.baxter,ou=shadow,o=Tremolo' filter='(objectClass=*)' scope='0' attribs=''
[2020-02-07 19:47:57,260][XNIO-1 task-10] INFO  AccessLog - RESULT op=323 con=322 result=0 time=9
[2020-02-07 19:47:57,278][XNIO-1 task-10] INFO  AccessLog - [AzSuccess] - k8sIdp - https://k8s-auth-prototype.corp.cira.ca/auth/idp/k8sIdp/token - uid=alain.baxter,ou=shadow,o=Tremolo - NONE [192.168.217.220] - [f89ac1341bb133231ac27de62a44cf94edf1a51b4]
[2020-02-07 19:47:57,280][XNIO-1 task-9] INFO  AccessLog - SRCH op=324 con=323 base='o=Tremolo' filter='(sub=alain.baxter)' scope='2' attribs=''
[2020-02-07 19:47:57,290][XNIO-1 task-9] INFO  AccessLog - RESULT op=324 con=323 result=0 time=10
[2020-02-07 19:47:57,291][XNIO-1 task-9] INFO  AccessLog - SRCH op=325 con=324 base='uid=alain.baxter,ou=shadow,o=Tremolo' filter='(objectClass=*)' scope='0' attribs=''
[2020-02-07 19:47:57,300][XNIO-1 task-9] INFO  AccessLog - RESULT op=325 con=324 result=0 time=9
[2020-02-07 19:47:57,302][XNIO-1 task-9] INFO  AccessLog - SRCH op=326 con=325 base='uid=alain.baxter,ou=shadow,o=Tremolo' filter='(objectClass=*)' scope='0' attribs=''
[2020-02-07 19:47:57,311][XNIO-1 task-9] INFO  AccessLog - RESULT op=326 con=325 result=0 time=9
[2020-02-07 19:47:57,314][XNIO-1 task-9] INFO  AccessLog - SRCH op=327 con=326 base='uid=alain.baxter,ou=shadow,o=Tremolo' filter='(objectClass=*)' scope='0' attribs=''
[2020-02-07 19:47:57,325][XNIO-1 task-9] INFO  AccessLog - RESULT op=327 con=326 result=0 time=11
[2020-02-07 19:47:57,327][XNIO-1 task-9] INFO  AccessLog - SRCH op=328 con=327 base='uid=alain.baxter,ou=shadow,o=Tremolo' filter='(objectClass=*)' scope='0' attribs=''
[2020-02-07 19:47:57,336][XNIO-1 task-9] INFO  AccessLog - RESULT op=328 con=327 result=0 time=9
[2020-02-07 19:47:57,352][XNIO-1 task-9] INFO  AccessLog - [AuSuccess] - k8s - https://k8s-dash-prototype.corp.cira.ca/auth/oidc - uid=alain.baxter,ou=shadow,o=Tremolo - 20 / k8s-db-oidc [192.168.217.220] - [f2dc6d646612015628081d8fe79bee4c108415e29]
[2020-02-07 19:47:57,364][XNIO-1 task-3] INFO  AccessLog - [AzSuccess] - k8s - https://k8s-dash-prototype.corp.cira.ca/ - uid=alain.baxter,ou=shadow,o=Tremolo - oauth2bearer [192.168.217.220] - [f2dc6d646612015628081d8fe79bee4c108415e29]
...
[2020-02-07 19:47:58,774][XNIO-1 task-3] INFO  AccessLog - [AzSuccess] - k8s - https://k8s-dash-prototype.corp.cira.ca/api/v1/namespace - uid=alain.baxter,ou=shadow,o=Tremolo - oauth2bearer [192.168.217.220] - [f2dc6d646612015628081d8fe79bee4c108415e29]
[2020-02-07 19:47:58,776][XNIO-1 task-2] INFO  AccessLog - [AzSuccess] - k8s - https://k8s-dash-prototype.corp.cira.ca/api/v1/cronjob/default - uid=alain.baxter,ou=shadow,o=Tremolo - oauth2bearer [192.168.217.220] - [f2dc6d646612015628081d8fe79bee4c108415e29]
[2020-02-07 19:47:58,792][XNIO-1 task-6] INFO  AccessLog - [AzSuccess] - k8s - https://k8s-dash-prototype.corp.cira.ca/api/v1/daemonset/default - uid=alain.baxter,ou=shadow,o=Tremolo - oauth2bearer [192.168.217.220] - [f2dc6d646612015628081d8fe79bee4c108415e29]
[2020-02-07 19:47:58,793][XNIO-1 task-8] INFO  AccessLog - [AzSuccess] - k8s - https://k8s-dash-prototype.corp.cira.ca/api/v1/deployment/default - uid=alain.baxter,ou=shadow,o=Tremolo - oauth2bearer [192.168.217.220] - [f2dc6d646612015628081d8fe79bee4c108415e29]
[2020-02-07 19:47:58,793][XNIO-1 task-4] INFO  AccessLog - [AzSuccess] - k8s - https://k8s-dash-prototype.corp.cira.ca/api/v1/job/default - uid=alain.baxter,ou=shadow,o=Tremolo - oauth2bearer [192.168.217.220] - [f2dc6d646612015628081d8fe79bee4c108415e29]
[2020-02-07 19:47:58,793][XNIO-1 task-9] INFO  AccessLog - [AzSuccess] - k8s - https://k8s-dash-prototype.corp.cira.ca/api/v1/pod/default - uid=alain.baxter,ou=shadow,o=Tremolo - oauth2bearer [192.168.217.220] - [f2dc6d646612015628081d8fe79bee4c108415e29]
```

So it does look like I'm coming through OpenUnison correctly.

From the same Session I am able to authenticate kubectl correctly

**Comment by mlbiam** [comment]
> I was inspecting the traffic using Chrome Developer Tools.

Ahh, you won't see it there.  OpenUnison is a reverse proxy injecting the token on each request.  So Chrome would never see the `Authorization` header being added to the request.

So you're right, its clearly going through OpenUnison.  Are you using a *nix? (ie Linux or mac as your client).  if so, try this:

1.  Open a new terminal
2.  `export KUBECONFIG=/tmp/openunison.config`
3.  Copy & paste the kubectl command into the terminal and run `kubectl get ns`

Did it work?

**Comment by alain-baxter** [comment]
Yup I'm using a *nix system for my client.

This did work.

```
[alain.baxter@dev-k8s-01 ~]$ cat /tmp/openunison.config
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUN5RENDQWJDZ0F3SUJBZ0lCQURBTkJna3Foa2lHOXcwQkFRc0ZBREFWTVJNd0VRWURWUVFERXdwcmRXSmwKY201bGRHVnpNQjRYRFRJd01ERXlPVEU0TkRNeE5sb1hEVE13TURFeU5qRTRORE14Tmxvd0ZURVRNQkVHQTFVRQpBeE1LYTNWaVpYSnVaWFJsY3pDQ0FTSXdEUVlKS29aSWh2Y05BUUVCQlFBRGdnRVBBRENDQVFvQ2dnRUJBTXlQCnNoSjRWZWJjbndCTnRMMitMekFKVGk0bzJvV0N0b1pzN05WZ2NpSXRVUkczbFE4VTdVTitoWjUwbFVrdkhaQlQKd3hqZVI1dHdwQnhlWXpiR29nNnJyelZJeGdXNlpLSENzd1ZyUU51YjdWa3lKd29PZ3d6WDM1V25RcWRIaVVLLwpvdmdrLzVsWVNVbUVjYm00UlVKSzdsU0c3QWtMVElucys0dDY4aldUc3JrRjJGaDIxZENLdXhpVHN5TUJ1S25RCkRheFl5SDFSRUlKdXZhSWIzcUo4U2RyTHBHYTV0Q2ZPTWVZc29yVzRkMm1hSjBkTVZFaGJVQUFNelkwdzZ2R0gKYTRzWHpZZUJHVlJNRWNOenhPUTNrVU9ZZC9LK090ekxtUU9EQjlBRWpaVXBpQkRZZjNvRFROZmNOM1NmYS8vcgpYbDA4dHkwcUQzOE5HOTdPWGc4Q0F3RUFBYU1qTUNFd0RnWURWUjBQQVFIL0JBUURBZ0trTUE4R0ExVWRFd0VCCi93UUZNQU1CQWY4d0RRWUpLb1pJaHZjTkFRRUxCUUFEZ2dFQkFEVzhYVkRmY1dNdWo0UWs2Vk5ZSFlmTVY3TFEKd09oTmdqTDRrTlBiUWUwVHRTZ0JpRFpkZXE3NVFxR0JrVVlpTHdKNk10bGZqSHd1RThqc0J1YWt0SXJUVDRKMgpiOVJtb2dsaGdDWDIxZ1I3NENnbUoyYWRDTTBqU0tWdWh2VWtVSlQxN005TEZPaXRudFFUbXVxNTJlMW5Ca0tMCjFoMTR4RFJkL25uc0dCejVJNXpQM1F3NlA2RWprNGhYSExSTVhWcCtnS2tDYjMyYms2UVRXUWpISFNHZVZaN0wKRHZuQ3Z5UGNWdEpXQS81UUc2K3NsRWxvLzNxMU1QcGJKK09BVHZIYlJsM0xzTkdCTFVablZpVCt1MFFLNjhsMApFN0V2Tm5nVUxGTS9zd3grMGRYZm1kYURHMkpFRFdCbmdCaERCSHNrdmFxdThQZCsrcmNLSVRrQWdFdz0KLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQo=
    server: https://dev-k8s-01.corp.cira.ca:6443
  name: kubernetes
contexts:
- context:
    cluster: kubernetes
    user: alain.baxter
  name: kubernetes
current-context: kubernetes
kind: Config
preferences: {}
users:
- name: alain.baxter
  user:
    auth-provider:
      config:
        client-id: kubernetes
        id-token: eyJraWQiOiJDTj11bmlzb24tc2FtbDItcnAtc2lnLCBPVT1LdWJlcm5ldGVzLCBPPUNhbmFkaWFuIEludGVybmV0IFJlZ2lzdHJhdGlvbiBBdXRob3JpdHksIEw9T3R0YXdhLCBTVD1PbnRhcmlvLCBDPUNBLUNOPXVuaXNvbi1zYW1sMi1ycC1zaWcsIE9VPUt1YmVybmV0ZXMsIE89Q2FuYWRpYW4gSW50ZXJuZXQgUmVnaXN0cmF0aW9uIEF1dGhvcml0eSwgTD1PdHRhd2EsIFNUPU9udGFyaW8sIEM9Q0EtMTU4MDg1MDUyNjEzMyIsImFsZyI6IlJTMjU2In0.eyJpc3MiOiJodHRwczovL2s4cy1hdXRoLXByb3RvdHlwZS5jb3JwLmNpcmEuY2EvYXV0aC9pZHAvazhzSWRwIiwiYXVkIjoia3ViZXJuZXRlcyIsImV4cCI6MTU4MTEwNDkyMSwianRpIjoid0ppUVhkYkNNRDhKZVlZeG1QNU1ldyIsImlhdCI6MTU4MTEwNDg2MSwibmJmIjoxNTgxMTA0NzQxLCJzdWIiOiJhbGFpbi5iYXh0ZXIiLCJuYW1lIjoiICIsImdyb3VwcyI6WyJDTj1BRFAsT1U9T2ZmaWNlIDM2NSBDb21tb24gTWFpbGJveCxPVT1DSVJBIEdyb3VwcyxEQz1DT1JQLERDPUNJUkEsREM9Q0EiLCJDTj1CYW1ib29IUixPVT1PZmZpY2UgMzY1IENvbW1vbiBNYWlsYm94LE9VPUNJUkEgR3JvdXBzLERDPUNPUlAsREM9Q0lSQSxEQz1DQSIsIkNOPUJlYXVjZXJvbixPVT1PZmZpY2UgMzY1IENvbW1vbiBNYWlsYm94LE9VPUNJUkEgR3JvdXBzLERDPUNPUlAsREM9Q0lSQSxEQz1DQSIsIkNOPUxpbmtlZEluLE9VPU9mZmljZSAzNjUgQ29tbW9uIE1haWxib3gsT1U9Q0lSQSBHcm91cHMsREM9Q09SUCxEQz1DSVJBLERDPUNBIiwiQ049U21hbGxJbXByb3ZlbWVudCxPVT1PZmZpY2UgMzY1IENvbW1vbiBNYWlsYm94LE9VPUNJUkEgR3JvdXBzLERDPUNPUlAsREM9Q0lSQSxEQz1DQSIsIkNOPWdvb2dsZWFjZWksT1U9UmVzdHJpY3RlZCxPVT1EaXN0cmlidXRpb24sT1U9VXNlciBHcm91cHMsT1U9Q0lSQSBHcm91cHMsREM9Q09SUCxEQz1DSVJBLERDPUNBIiwiQ049QXp1cmUtU3Vic2NyaXB0aW9uLUNvQWRtaW5pc3RyYXRvcixPVT1PZmZpY2UgMzY1IENvbW1vbiBNYWlsYm94LE9VPUNJUkEgR3JvdXBzLERDPUNPUlAsREM9Q0lSQSxEQz1DQSIsIkNOPUxpbnV4LURldk9wcyxPVT1ncm91cHMsT1U9TGludXgsREM9Q09SUCxEQz1DSVJBLERDPUNBIiwiQ049c3J2X1dpbi1UZWFtQ2l0eSxPVT1TZXJ2ZXIgQWNjZXNzLE9VPVVzZXIgR3JvdXBzLE9VPUNJUkEgR3JvdXBzLERDPUNPUlAsREM9Q0lSQSxEQz1DQSIsIkNOPVZhdWx0IEJpeiBVc2VycyxPVT1WYXVsdCxPVT1TZXJ2aWNlIEFjY291bnRzLERDPUNPUlAsREM9Q0lSQSxEQz1DQSIsIkNOPXNlY19RQSxPVT1TZWN1cml0eSxPVT1Vc2VyIEdyb3VwcyxPVT1DSVJBIEdyb3VwcyxEQz1DT1JQLERDPUNJUkEsREM9Q0EiLCJDTj1GdXJ5IEwzIFN1cHBvcnQsT1U9RS1NYWlsIERpc3RyaWJ1dGlvbiBHcm91cHMsT1U9R2Fsc3luY0NPUlAsREM9Q09SUCxEQz1DSVJBLERDPUNBIiwiQ049UUEsT1U9UUEsT1U9R2Fsc3luY0NPUlAsREM9Q09SUCxEQz1DSVJBLERDPUNBIiwiQ049TGludXgtUUEsT1U9Z3JvdXBzLE9VPUxpbnV4LERDPUNPUlAsREM9Q0lSQSxEQz1DQSIsIkNOPUNJUkEgU1RBRkYsQ049VXNlcnMsREM9Q09SUCxEQz1DSVJBLERDPUNBIl0sInByZWZlcnJlZF91c2VybmFtZSI6ImFsYWluLmJheHRlciJ9.SYFLtTXeHemDIlB4QkAeHh9YfyddwknfXMPm-NIe6FG1ROjTErFRFvBzCFAlIqNPu-wZOJc_0DcFIRcK0gO1s00oGP0qwNhTAYWK441n8MCR5WegXLF62QE-dnPnfyrvNOno-ovmiS_ny_T2b0ba1GLIyOfnDSV4zmKk6O26lkJh4c9gxabqIjJnSQXoMEvIdR8d6BNJa1Fdfzn48m7K8QYyNnHOstxeBlpQbROXdDwBEp2lbTJoJ5pgiCPrAJt0UyIrcjGoYJ7Ay26oaSeRKclQZKh1Vdqc2DlN6AxrUn89KEum-GrjxLFVScJwSQCAOwuCNJv0LfEO1ZwVT-jqmg
        idp-certificate-authority-data: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUVURENDQXpTZ0F3SUJBZ0lHQVhBU0NMYllNQTBHQ1NxR1NJYjNEUUVCQ3dVQU1JR2lNUXN3Q1FZRFZRUUcNCkV3SkRRVEVRTUE0R0ExVUVDQk1IVDI1MFlYSnBiekVQTUEwR0ExVUVCeE1HVDNSMFlYZGhNVEV3THdZRFZRUUsNCkV5aERZVzVoWkdsaGJpQkpiblJsY201bGRDQlNaV2RwYzNSeVlYUnBiMjRnUVhWMGFHOXlhWFI1TVJNd0VRWUQNClZRUUxFd3BMZFdKbGNtNWxkR1Z6TVNnd0pnWURWUVFERXg5ck9ITXRZWFYwYUMxd2NtOTBiM1I1Y0dVdVkyOXkNCmNDNWphWEpoTG1OaE1CNFhEVEl3TURJd05ESXhNRGcwTlZvWERUSXhNREl3TXpJeE1EZzBOVm93Z2FJeEN6QUoNCkJnTlZCQVlUQWtOQk1SQXdEZ1lEVlFRSUV3ZFBiblJoY21sdk1ROHdEUVlEVlFRSEV3WlBkSFJoZDJFeE1UQXYNCkJnTlZCQW9US0VOaGJtRmthV0Z1SUVsdWRHVnlibVYwSUZKbFoybHpkSEpoZEdsdmJpQkJkWFJvYjNKcGRIa3gNCkV6QVJCZ05WQkFzVENrdDFZbVZ5Ym1WMFpYTXhLREFtQmdOVkJBTVRIMnM0Y3kxaGRYUm9MWEJ5YjNSdmRIbHcNClpTNWpiM0p3TG1OcGNtRXVZMkV3Z2dFaU1BMEdDU3FHU0liM0RRRUJBUVVBQTRJQkR3QXdnZ0VLQW9JQkFRQ2YNCjJGc0cvQkdYTExwUzMzbWg5akY5VnhVeDVwMGJ4b0h5T29mNCtyZ0tac0tUYzJpb2lLNEFTL1NHVGNubW9rZlgNClRaOWxURGN5ckd0aU1wRTBDbU9HOUdITHMrZzhLT0RGKy9KWUs4Y0F5cVNaaW80WnlsRGlaQXZmc2ovS2cwZUYNCkxQWFJUSnBGdnQ4S0xieVYxOWgycjhBQTR4cVRhb1BFM1paQ2lWSmVHeUEvdlZLekJNdzNsc2RBUFVCeCtVNmYNCm9GSy9NWDV2VTJHSEFxV09tVnZScU1veFkvcW9uaHF1dWdmUWFucytWbWNrWk03TlQ2c3Z3RnVkN3dqVGRvY2wNCnhvdVFaQTU0RzIwWlZ6bTl6eHJnTE9xL0hONEllaUV1Y3hBWVlXSzdpZnhhNlkyR1pUeGlGNGpxNTViWTBSR2wNClcwQWlEaWROYVdocWFCcFFBWGR4QWdNQkFBR2pnWVV3Z1lJd0R3WURWUjBUQVFIL0JBVXdBd0VCL3pBT0JnTlYNCkhROEJBZjhFQkFNQ0FnUXdFZ1lEVlIwbEFRSC9CQWd3QmdZRVZSMGxBREJMQmdOVkhSRUVSREJDZ2g5ck9ITXQNCllYVjBhQzF3Y205MGIzUjVjR1V1WTI5eWNDNWphWEpoTG1OaGdoOXJPSE10WkdGemFDMXdjbTkwYjNSNWNHVXUNClkyOXljQzVqYVhKaExtTmhNQTBHQ1NxR1NJYjNEUUVCQ3dVQUE0SUJBUUNlaFJNOTZYZHJzWW1YZzJIZzF5ek0NClRQRmR6S0VHRGRvWmdXSFMzeDlOTXF1Unl2dHBCSGUvdFBGU0ZBQzVvc3RGeExMOUM3NDA5cEVnaXF0OWVHVHgNCkh1WFRSZTdJWHA1eEtYTkZibm11WFdBMEc0eVRBOEthN2RnbGtEUTFmWi85ZW5ZYjJZUERaTEZLeURweUF5Q3ANCkRYdUdSeWxBWjhrbi9wN3dNdkd6Q0kxSkIxcVVGUDdCUmdQWmlUWmlqQjVTTmJMTmJMdzM3b2RzUlVnc1RldHINClRMNlcwTXlGa2lMTExzUkJ3WTY3b0V4UCttWGF0WEp2RE9CQTNUU2pjM0lMTXFSZVpiNjB4T1RJWFRVa2hhWFoNCnFGdzRHM2pIVldkbTlKejBWRU84WjNGeTlpZGQwVm9mSk42TWh6a0xTZ2gwU0RxUTdwWm1VaFY1eGMwNU9xY2INCi0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0=
        idp-issuer-url: https://k8s-auth-prototype.corp.cira.ca/auth/idp/k8sIdp
        refresh-token: q1bKLFOyUvL29g+PcjLLishxC8lMdzbUz8ozMXGMKTUwME6BkEo6Sql5yUWVBSWpKUGphaWpxSVAfSHBJeFFxdmVxf6ZLqHFZvnOFuWFFsHuvokFKYlGUR6WBQa+uZ7maabG5Y5+lTkeVUEWbh5ARZHFUaWhXpmZRUq1AA==
      name: oidc
[alain.baxter@dev-k8s-01 ~]$ kubectl get ns
NAME                   STATUS   AGE
default                Active   9d
ingress-nginx          Active   7d4h
kube-node-lease        Active   9d
kube-public            Active   9d
kube-system            Active   9d
kubernetes-dashboard   Active   7d3h
openunison             Active   2d22h
openunison-deploy      Active   2d22h
```

**Comment by mlbiam** [comment]
that is really odd.  when you access the dashboard, does the API server generate any logs like "invalid token"?  Anything related to oidc?

**Comment by alain-baxter** [comment]
I didn't see anything in the kube-apiserver logs.

All I see in the dashboard logs is:
```
2020/02/07 20:11:34 [2020-02-07T20:11:34Z] Incoming HTTP/1.1 GET /api/v1/namespace request from 10.6.20.34:52758:
2020/02/07 20:11:34 Getting list of namespaces
2020/02/07 20:11:34 Non-critical error occurred during resource retrieval: Unauthorized
2020/02/07 20:11:34 [2020-02-07T20:11:34Z] Outcoming response to 10.6.20.34:52758 with 200 status code
```

**Comment by mlbiam** [comment]
@alain-baxter i deleted your comment on it working because you had embedded both your idtoken and refresh token.  The idtoken has PII in it that is easily decodable and while the token its self is only good for a couple of minutes with the refresh token it could be compromised.

**Comment by alain-baxter** [comment]
@mlbiam Thanks

**Comment by mlbiam** [comment]
this is so odd.  ok, lets break the api server and make sure that kubectl stops working.  (kubectl and dashboard are using keys encoded with the same certificate).  In your api server manifest (I'm assuming kubeadm?) comment out the flag for the ca file.  Then copy and paste your kubectl command and run `kubectl get ns`.

**Comment by alain-baxter** [comment]
Okay, i've commented out the ca file flag, waitted for the API server to come back up and tested kubectl, I'm now unauthorized.
```
[alain.baxter@dev-k8s-01 ~]$ kubectl get ns
error: You must be logged in to the server (Unauthorized)
```

**Comment by mlbiam** [comment]
ok, so fix the api server so you're back where you started (with kubectl working and the dashboard not working).  Do the unauthorized errors just say "Unauthorized" on the screen or does it say something like "User https://sdfsdfsdfsdf:user is not authorized"...?

**Comment by alain-baxter** [comment]
![unauthorized-dashboard](https://user-images.githubusercontent.com/21297900/74063886-73498280-49bf-11ea-8a7d-68c76977f496.PNG)

Just says "Unauthorized" in the notifications

**Comment by alain-baxter** [comment]
![user-dropdown](https://user-images.githubusercontent.com/21297900/74064350-7d1fb580-49c0-11ea-91f4-721c3631c4a4.PNG)

When I dropdown the User Icon is says "Logged in with Auth Header"

**Comment by mlbiam** [comment]
OK, so it sees the header.  Check the clocks on your master and nodes.  are they connected to an NTP server and correct?

**Comment by alain-baxter** [comment]
```
[alain.baxter@dev-k8s-01 ~]$ sudo clock
Fri 07 Feb 2020 08:49:30 PM UTC  -0.661404 seconds
```
looks correct, we run our systems at UTC vs local time

**Comment by mlbiam** [comment]
is that 12 hours off?  If you're in Canada wouldn't it be Sat Feb 8 since we're behind UTC in this hemosphere or am I reading this wrong?

**Comment by alain-baxter** [comment]
I'm in EST which is UTC-5, so this is the correct time

**Comment by mlbiam** [comment]
oh, i see my mistake.  the PM.  use to uts in 24 hour clock

**Comment by alain-baxter** [comment]
I can tell from the logs that OpenUnison is also running UTC, which is also the correct time

**Comment by mlbiam** [comment]
OK, so lets play what do we know:
1.  API server is configured correctly
2.  The kubectl command works
3.  The dashboard is receiving the header
4.  The clocks on the nodes and masters are all in-sync
5.  There are no error messages in the api server
6.  The api server is rejecting the token, not just the authorization of the user

Did I miss any pertinent facts?  A couple of other things to check:

1.  take your `id_token` over to jwt.io (nothing gets sent back home), is there a `sub` claim?
2.  What distro of k8s are you using?

**Comment by alain-baxter** [comment]
I've taken my id_token to jwt.io, it does have a `sub` claim:
```
{
  "iss": "https://k8s-auth-prototype.corp.cira.ca/auth/idp/k8sIdp",
  "aud": "kubernetes",
  "exp": 1581360000,
  "jti": "vm6-mMuZuRwU-2tkoIO7hQ",
  "iat": 1581359940,
  "nbf": 1581359820,
  "sub": "alain.baxter",
  "name": " ",
  "groups": [
    <snip>
  ],
  "preferred_username": "alain.baxter"
}
```

As for Distro, I'm working on an on premise cluster that was built using kubeadm. Very basic, the core install + CNI (Calico).

I've added Nginx Ingress and the Dashboard just for testing out OpenUnison, just doing the recommended install of each.

**Comment by mlbiam** [comment]
Lets try seeing if the dashboard likes the idtoken from kubectl:

1.  Create an ingress based on the below yaml.  Replace `k8dbdirect.apps.192-168-159-168.nip.io` with the host name you want to use with your direct dashboard access

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/affinity: cookie
    nginx.ingress.kubernetes.io/backend-protocol: https
    nginx.ingress.kubernetes.io/secure-backends: "true"
    nginx.ingress.kubernetes.io/session-cookie-hash: sha1
    nginx.ingress.kubernetes.io/session-cookie-name: kubernetes-dashboard
    nginx.org/ssl-services: kubernetes-dashboard
  name: kubernetes-dashboard
  namespace: kubernetes-dashboard
spec:
  rules:
  - host: k8dbdirect.apps.192-168-159-168.nip.io
    http:
      paths:
      - backend:
          serviceName: kubernetes-dashboard
          servicePort: 443
        path: /
  
  tls:
  - hosts:
    - k8sdb.apps.192-168-159-168.nip.io
```

2.  Once deployed, go to the direct ingress, youll be asked your kubeconfig or your token.  Choose token
3.  In another tab, login to openunison and click on "Kubernetes Token" just like if you are getting your kubectl token
4. Click on the copy button next to `id_token`
5.  Paste this token into the direct dashboard access and see what happens

Once 5 is done the dashboard should be usable for a couple of minutes.

**Comment by alain-baxter** [comment]
I have tried the steps provided, and I think it uncovers the main reason.

I got the error: `Unauthorized (401): Invalid credentials provided`

So the dashboard doesn't seem to be able to verify my credentials.

**Comment by alain-baxter** [comment]
To install the dashboard, I ran 1 command:

`kubectl apply -f https://raw.githubusercontent.com/kubernetes/dashboard/v2.0.0-beta8/aio/deploy/recommended.yaml`

**Comment by alain-baxter** [comment]
In the dashboard logs I see the following:
```
2020/02/11 19:53:24 [2020-02-11T19:53:24Z] Incoming HTTP/1.1 GET /api/v1/csrftoken/login request from 192.168.217.203:45668: { contents hidden }
2020/02/11 19:53:24 [2020-02-11T19:53:24Z] Outcoming response to 192.168.217.203:45668 with 200 status code
2020/02/11 19:53:24 [2020-02-11T19:53:24Z] Incoming HTTP/1.1 POST /api/v1/login request from 192.168.217.203:45668: { contents hidden }
2020/02/11 19:53:24 Non-critical error occurred during resource retrieval: the server has asked for the client to provide credentials
2020/02/11 19:53:24 [2020-02-11T19:53:24Z] Outcoming response to 192.168.217.203:45668 with 200 status code
```

**Comment by mlbiam** [comment]
thats odd.  the dashboard doesn't verify anything.  it just passes it on to the api server.  You mentioned that you added the nginx ingress because thats what openunison requires.  Is there another ingress on your cluster?

**Comment by alain-baxter** [comment]
We don't have another Ingress tooling yet for the cluster

**Comment by mlbiam** [comment]
hmm.  OK, lets try this:

1.  Login to openunison, generate kubectl and run `kubectl get ns` (this will force a refresh if needed)
2.  try `curl -v --insecure https://apihost:port/api/v1/namespaces` You should get something like:
```
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "namespaces is forbidden: User \"system:anonymous\" cannot list resource \"namespaces\" in API group \"\" at the cluster scope",
  "reason": "Forbidden",
  "details": {
    "kind": "namespaces"
  },
  "code": 403
* Connection #0 to host 192.168.159.168 left intact
}* Closing connection 0
```
3.  Run `k get ns` again to refresh the token if needed
4.  Run `export ID_TOKEN=$(grep id-token ~/.kube/config | awk '{print $2}')`
5.  Run `curl -v --insecure -H "Authorization: Bearer $ID_TOKEN" https://apihost:port/api/v1/namespaces`
6.  You should see a bunch of namespaces
7.  Launch the ingress direct and choose token
8.  run `kubectl get ns` to refresh the token
9.  run `export ID_TOKEN=$(grep id-token ~/.kube/config | awk '{print $2}')`
10. run `echo $ID_TOKEN`
11.  Take the output and put it in is the token for dashboard

Did it work?  

Sorry for getting so detailed on the test but this is really kinda bonkers.  The only reason teh dashboard wouldn't pass the token through is if the dashboard isn't running on TLS (which it is, otherwise openunison couldn't talk to it)

**Comment by alain-baxter** [comment]
@mlbiam I think I solved it this morning, it was on me.

I have 3 master nodes setup, but not a load balancer for them yet. So my kubectl requests were going directly to one of the api servers.

I had copied the `ou-ca.pem` to all 3 masters, but I forgot the oidc flags on the other 2. The dashboard uses the internal svc for the apiserver, which means it will load balance to all 3 of them. Oddly I wasn't getting 66% failure on requests, but 100% from the dashboard, maybe it was favoring the 2 unused master apiservers since the 1 was getting all the cluster traffic from kubelet.

I've configured the oidc flags on the other 2 and the Dashboard authentication is working.
